platform/sync-installer/sync-installer-impl/src/main/java/org/codice/ddf/sync/installer/impl/SynchronisedInstallerImpl.java

```diff
@@ -0,0 +1,566 @@
+          DEFAULT_POLLING_INTERVAL,
+          "Managed service failed to appear after [" + maxWaitTime + "] ms.");
+    } finally {
+      st.close();
+    }
+  }
+
+  @Override
+  public void installFeatures(String feature, String... additionalFeatures)
+      throws SynchronisedInstallerException {
+    installFeatures(DEFAULT_MAX_FEATURE_WAIT, feature, additionalFeatures);
+  }
+
+  @Override
+  public void installFeatures(long maxWaitTime, String feature, String... additionalFeatures)
+      throws SynchronisedInstallerException {
+
+    Set<String> featuresToInstall =
+        getFeaturesFromNames(feature, additionalFeatures)
+            .filter(f -> !featuresService.isInstalled(f))
+            .map(Feature::getName)
+            .collect(Collectors.toSet());
+
+    LOGGER.info("Installing the following features: [{}]", String.join(", ", featuresToInstall));
+    long startTime = System.currentTimeMillis();
+    try {
+      AccessController.doPrivileged(
+          (PrivilegedExceptionAction<Void>)
+              () -> {
+                featuresService.installFeatures(
+                    featuresToInstall, EnumSet.of(FeaturesService.Option.NoAutoRefreshBundles));
+                return null;
+              });
+    } catch (Exception e) {
+      throw new SynchronisedInstallerException(
+          "Failed to install features [" + String.join(",", featuresToInstall) + "]", e);
+    }
+
+    waitForBundles(maxWaitTime);
+    LOGGER.info("Finished installing features in {} ms", (System.currentTimeMillis() - startTime));
+  }
+
+  @Override
+  public void uninstallFeatures(String feature, String... additionalFeatures)
+      throws SynchronisedInstallerException {
+    uninstallFeatures(DEFAULT_MAX_FEATURE_WAIT, feature, additionalFeatures);
+  }
+
+  @Override
+  public void uninstallFeatures(long maxWaitTime, String feature, String... additionalFeatures)
+      throws SynchronisedInstallerException {
+    Set<String> featuresToUninstall =
+        getFeaturesFromNames(feature, additionalFeatures)
+            .filter(f -> featuresService.isInstalled(f))
+            .map(Feature::getName)
+            .collect(Collectors.toSet());
+
+    LOGGER.info(
+        "Uninstalling the following features: [{}]", String.join(", ", featuresToUninstall));
+    long startTime = System.currentTimeMillis();
+    try {
+      AccessController.doPrivileged(
+          (PrivilegedExceptionAction<Void>)
+              () -> {
+                featuresService.uninstallFeatures(
+                    featuresToUninstall, EnumSet.of(FeaturesService.Option.NoAutoRefreshBundles));
+                return null;
+              });
+
+    } catch (Exception e) {
+      throw new SynchronisedInstallerException(
+          "Failed to uninstall features [" + String.join(",", featuresToUninstall) + "]", e);
+    }
+
+    waitForBundles(maxWaitTime);
+    LOGGER.info(
+        "Finished uninstalling features in {} ms", (System.currentTimeMillis() - startTime));
```

Comment:
✏️ Might want to trace which features?

Reply:
I am simply asking to list which features otherwise in a multi-threaded system, it won't be obvious which one goes with which start trace.